File: lib/datainterp.py
#Author: Zhengyi Yang
#Created on Fri 20 Nov, 2015

from numpy import array,float16,meshgrid,isnan
from scipy.interpolate import CloughTocher2DInterpolator
import pickle

def gen_func(x,y,z):
    #ziped_xyz=zip(zip(x,y),z)
    #ziped_xyz=[x for x in ziped_xyz if x[1]!=0]
    #unziped_xyz=zip(*ziped_xyz)
    #points=array(unziped_xyz[0],dtype=float64)
    #z=array(unziped_xyz[1],dtype=float64)
    points=array(zip(x,y),dtype=float16)
    z=array(z,dtype=float16)

    # print points
    # print z
        
    f = CloughTocher2DInterpolator(points,z)#Clough-Tocher method
    with open('estimated_func.pkl', 'wb') as output:
        pickle.dump(f, output, pickle.HIGHEST_PROTOCOL)
    
    #print f(x,y)
    #make_graph(x,y,z)

def load_func(x,y,isCoord=False):
    with open('estimated_func.pkl', 'rb') as input:
        f = pickle.load(input)
    x=array(x,dtype=float16)
    y=array(y,dtype=float16)
    if isCoord:
        xx,yy=meshgrid(x,y)
        ans=f(xx,yy)
    else:
        ans=f(x,y)
    ans=ans.tolist()
    return ans

File: lib/datainterp.rb
#Author: Zhengyi Yang
#Created on Fri 20 Nov, 2015

require 'singleton'
require 'rubypython'
require 'pathname'

class DataInterp
#This class performs a 2-dimensional Clough-Tocher interpolation on a given dataset.
#The interpolation is supported by Python(version 2.x) SciPy library and a Ruby-Python 
#bridge is built using rubypython.
#Please try the following commands to install these libraries: 
# $: pip install numpy
# $: pip install scipy
# $: gem install rubypython
#Example:
# interp=DataInterp.new(x,y,z)
# aqi=interp.clac(i,j)
# interp.stop

  include Singleton

  def initialize()
  #This method creates a new DataInterp instance.
  #longs -- an array of longitudes of all data points
  #lats -- an array of latitudes of all data points
  #aqis -- an array of AQIs(pollution levels) of all data points
    @isRunning=false
    pollution = Pollution.all
    if(pollution==nil)
      warn('warning: initializing DataInterp with no parameters, latest function will be loaded.')
      start
      return
    end
    start
    update(pollution)
  end
  
  def update(pollution)#Update all data points
    longitudes = pollution.map {|p| p.longitude}
    latitudes = pollution.map {|p| p.latitude}
    indexs = pollution.map {|p| p.index}
    @interp.gen_func(longitudes, latitudes, indexs)
  end

  def calc(longs,lats,isCoord=false)
  #This method computes estimated AQIs of given locations.
  #longs -- an array of longitudes of given locations.
  #lats -- an array of latitudes of given locations.
  #isCoord --true if longs and lats are axies
  #return -- an array of AQIs of given locations.
    raise RuntimeError,"Python interperter has been stopped" unless @isRunning
    raise ArgumentError, "Arguments must have the same size" unless longs.size==lats.size
    aqis=@interp.load_func(longs,lats,isCoord).rubify
    return aqis
  end
  
  def stop
  #Stop external Python interperter
  #Remember to stop Python interperter if the interpolation is no longer needed
  #Due to high likelihood of segementation fault, you cannot restart the Python
  #interperter anagin after stopping it.
    flag=RubyPython.stop
    raise RuntimeError,"Fail to stop python interperter" unless flag
    @isRunning=false
  end

  private
  def get_current_path
    return Pathname.new(File.dirname(__FILE__)).realpath.to_s
  end
  
  def start
  #Start external Python interperter
  #Note that this method is private to prevent segementation fault.
    flag=RubyPython.start
    raise RuntimeError,"Fail to start python interperter" unless flag
    @isRunning=true
    sys=RubyPython.import('sys')
    sys.path.append(get_current_path)
    os=RubyPython.import('os')
    os.chdir(get_current_path)
    @interp=RubyPython.import('datainterp')
  end
end

